File: clients/cs/Renderer.cs
using Raylib_cs;
using System.Diagnostics;

namespace cs_snake
{
    public class Renderer
    {
        private bool _shouldClose;
        private float _deltaTime;
        private int _cellSize;
        private int _width;
        private int _height;
        private Core _core;

        private float _timeSinceCoreUpdate;
        private float _coreUpdateFrequency;

        public float DeltaTime => _deltaTime;
        public bool ShouldClose => _shouldClose;
        public int Width => _width;
        public int Height => _height;
        public int CellSize => _cellSize;

        public Renderer(string title, Core core, int cellSize)
        {
            _core = core;
            (int width, int height) = core.GetGridDimensions();
            Debug.Assert(width > 0);
            Debug.Assert(height > 0);
            Debug.Assert(cellSize > 0);
            _deltaTime = 0.0f;
            _cellSize = cellSize;
            _width = width * cellSize;
            _height = height * cellSize;
            Raylib.InitWindow(_width, _height, title);
            _timeSinceCoreUpdate = 0;
            _coreUpdateFrequency = 1;
        }

        public void Destroy()
        {
            Raylib.CloseWindow();
        }

        public void Update()
        {
            if (!_core.CurrentState.isRunning)
            {
                return;
            }

            if(Raylib.IsKeyDown(KeyboardKey.W) || Raylib.IsKeyDown(KeyboardKey.Up))
            {
                _core.ChangeDirection(Direction.Up);
            }
            if(Raylib.IsKeyDown(KeyboardKey.D) || Raylib.IsKeyDown(KeyboardKey.Right))
            {
                _core.ChangeDirection(Direction.Right);
            }
            if(Raylib.IsKeyDown(KeyboardKey.S) || Raylib.IsKeyDown(KeyboardKey.Down))
            {
                _core.ChangeDirection(Direction.Down);
            }
            if(Raylib.IsKeyDown(KeyboardKey.A) || Raylib.IsKeyDown(KeyboardKey.Left))
            {
                _core.ChangeDirection(Direction.Left);
            }

            if (_timeSinceCoreUpdate > _coreUpdateFrequency)
            {
                _core.Update();
                _timeSinceCoreUpdate = 0;
            }
            _timeSinceCoreUpdate += Raylib.GetFrameTime();
        }

        public void Draw()
        {
            _shouldClose = Raylib.WindowShouldClose();

            Game game = Game.GetInstance();

            if (!_shouldClose)
            {
                Raylib.BeginDrawing();
                Raylib.ClearBackground(Raylib_cs.Color.Purple);

                GameState gameState = _core.CurrentState;

                foreach (var segment in gameState.segmentData)
                {
                    int x = segment.x * _cellSize;
                    int y = segment.y * _cellSize;
                    Raylib.DrawRectangle(x, y, _cellSize, _cellSize, Utils.GetRenderColor(segment.color));
                }

                if (!_core.CurrentState.isRunning)
                {
                    string text = "Game Over!";
                    int fontSize = 32;
                    int textWidth = Raylib.MeasureText(text, fontSize);
                    Raylib.DrawText(
                        text, 
                        _width / 2 - textWidth / 2, 0, 
                        fontSize,
                        Raylib_cs.Color.White);
                }

                Raylib.EndDrawing();
            }

            _deltaTime = Raylib.GetFrameTime();
        }
    }
}

File: clients/ts/src/renderer.ts
import type { Core } from "./core";
import { Direction, Utils } from "./utils";

export class Renderer {
  core: Core;
  cellSize: number;
  width: number;
  height: number;
  canvas: CanvasRenderingContext2D;
  lastFrameTime: number;
  deltaTime: number;
  timeSinceCoreUpdate: number;
  coreUpdateFrequency: number;
  pendingDirection: Direction | null;
  constructor(core: Core, cellSize: number) {
    this.core = core;
    this.cellSize = cellSize;
    const { width: gridWidth, height: gridHeight } = core.getGridDimensions();
    this.width = gridWidth * cellSize;
    this.height = gridHeight * cellSize;

    // Initialize canvas
    const canvasEl: HTMLCanvasElement =
      document.querySelector("#windowcanvas")!;
    const loadingEl: HTMLElement = document.querySelector(".windowloading")!;
    canvasEl.width = this.width;
    canvasEl.height = this.height;
    loadingEl.classList.remove("show");
    canvasEl.classList.add("show");
    this.canvas = canvasEl.getContext("2d")!;

    this.lastFrameTime = 0;
    this.deltaTime = 0;
    this.timeSinceCoreUpdate = 0;
    this.coreUpdateFrequency = 1000;
    this.pendingDirection = null;

    window.addEventListener("keydown", (event) => {
      if (event.key == "ArrowUp" || event.key == "w") {
        event.preventDefault();
        this.pendingDirection = Direction.Up;
      }
      if (event.key == "ArrowRight" || event.key == "d") {
        event.preventDefault();
        this.pendingDirection = Direction.Right;
      }
      if (event.key == "ArrowDown" || event.key == "s") {
        event.preventDefault();
        this.pendingDirection = Direction.Down;
      }
      if (event.key == "ArrowLeft" || event.key == "a") {
        event.preventDefault();
        this.pendingDirection = Direction.Left;
      }
    });
  }

  destroy() {}

  update() {
    this.deltaTime = Date.now() - this.lastFrameTime;
    if (this.deltaTime > 16) this.deltaTime = 16;
    this.lastFrameTime = Date.now();

    if (this.core.currentState.isRunning == false) return;

    this.timeSinceCoreUpdate += this.deltaTime;
    if (this.timeSinceCoreUpdate > this.coreUpdateFrequency) {
      if (this.pendingDirection != null) {
        this.core.changeDirection(this.pendingDirection);
        this.pendingDirection = null;
      }
      this.core.update();
      this.timeSinceCoreUpdate = 0;
    }
  }

  draw() {
    // Clear
    this.canvas.fillStyle = "purple";
    this.canvas.fillRect(0, 0, this.width, this.height);

    const gameState = this.core.currentState;

    gameState.segmentData.forEach((segment) => {
      this.canvas.fillStyle = Utils.getRenderColor(segment.color);
      this.canvas.fillRect(
        segment.x * this.cellSize,
        segment.y * this.cellSize,
        this.cellSize,
        this.cellSize,
      );
    });

    if (gameState.isRunning == false) {
      this.canvas.fillStyle = "white";
      const gameOverText = "Game Over!";
      this.canvas.font = "32px Arial";
      const { width: textWidth } = this.canvas.measureText(gameOverText);
      this.canvas.fillText(gameOverText, this.width / 2 - textWidth / 2, 50);
    }
  }
}
